### PR TITLE
doc: document python build scripts

### PR DESCRIPTION
**Pull request by dbkinder**
We have a collection of python scripts that are part of our build
system.  This PR collects docstring comments added to these scripts into
a summary document.  Previous references to just the script name in
other documentation are updated to point to this build tool
documentation.

Some of the scripts needed an update to be processed (via include
directives) consistently.

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

### CODE REVIEW

File: doc/guides/build_tools.rst [hunk 1]
```diff
@@ -0,0 +1,107 @@
+.. _build_tools:
+
+Build tools
```
Comment (nashif) [inline doc/guides/build_tools.rst]: Those are not build tools and having a guide called build tools is misleading. Maybe it should be called the "Zephyr Build Process" and provide more details about how those are invoked and when. We should not give the impression that a developer needs to invoke them.

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/guides/build_tools.rst (hunk 1 above)
Comment (dbkinder) [reply doc/guides/build_tools.rst]: How about "Internal build tools"?  @andrewboie can you give me a better intro paragraph?

**Review by marc-hb — COMMENTED** [review]
@dbkinder can you provide a convenience shortcut to the documentation build of this PR? Github's "Details" button doesn't lead to that.

BTW this looks like a good opportunity to add @carlescufi 's definition of `scripts/`
https://github.com/zephyrproject-rtos/zephyr/pull/15704#issuecomment-489278598

File: doc/guides/build_tools.rst (hunk 1 above)
Comment (nashif) [reply doc/guides/build_tools.rst]: > Considering they're "internal" and never invoked directly, do their pydocs even need to be published on the web site?
> 
> By the way one section of the documentation is currently called "User and Developer guides" but it doesn't seem to draw a line between a "User" and a "Developer". I haven't seen any line either between 1. an app developer (who shouldn't care about these scripts, black box) versus 2. a core Zephyr developer who may find useful gory implementation details about these scripts or about cmake (PR #13805).

let's not digress here. The documentation is needed, those however are neither "external" nor "internal" build tools. We need the documentation to be about the build process and what steps we take to generate the final binary, so there is a leading section that is missing here and probably some workflow diagram describing the build process. 
AFAIK we have this information and documentation, we just need to create some content around the text coming from the scripts.

File: doc/guides/build_tools.rst [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+:zephyr_file:`arch/x86/gen_idt.py`
+**********************************
+
+.. include:: ../../arch/x86/gen_idt.py
+   :start-after: """
+   :end-before: """
+
+.. _gen_gdt.py:
+
+:zephyr_file:`arch/x86/gen_gdt.py`
```
Comment (andrewboie) [inline doc/guides/build_tools.rst]: We may want to consider putting arch-specific scripts in their own area.

File: doc/guides/build_tools.rst [hunk 3]
```diff
@@ -0,0 +1,107 @@
+.. _build_tools:
+
+Build tools
+###########
+
+Zephyr project uses a collection of build and post-build scripts to
+assist creation of application images, as described throughout the
+Zephyr documentation, and summarized here.
+
```
Comment (andrewboie) [inline doc/guides/build_tools.rst]: missing IRQ scripts in arch/common

Comment (dbkinder) [reply doc/guides/build_tools.rst]: Thanks.  @andrewboie Can you add a docstring description to ``arch/common/gen_isr_tables.py``

Comment (andrewboie) [reply doc/guides/build_tools.rst]: will do

**Comment by dbkinder** [comment]
@marc-hb https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/16535/docs/guides/build_tools.html

**Comment by marc-hb** [comment]
> @marc-hb https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/16535/docs/guides/build_tools.html

Thanks! Now I can see. @nashif could this linking be automated and if yes is there already a ticket for it? If not I'll file it. If nothing else it'll be a place to share the format of the magic URLs.

So from a doc structure perspective,
- the higher-level  (and very good) **introductions** are somewhat hidden under "API **Reference** » User Mode", while
- these corresponding, relatively obscure scripts and low-level **implementation details** are being given higher visibility at  "User and Developer **Guides** » Build tools". 

If I'm new to something, I'm going to read "Guides" before any "(API) reference". I may never read "Reference" section actually, rather just search them.

> We need the documentation to be about the build process and what steps we take to generate the final binary, so there is a leading section that is missing here and probably some workflow diagram describing the build process.

Agreed, a bird's eye view of the build process and flow would be awesome. One of the reasons I started the #13805 "experiment" is because I missed that high-level view. However: 1. there's no such section about this yet AFAIK, 2. I'm still not sure it would be a better place for specific implementation details details like these scripts. Maybe sometimes but in this case I think having everything user mode in the "User Mode" section makes more sense. The number of "long links/jumps" crossing all that distance is IMHO a sign they should just be together.

**Review by nashif — APPROVED** [review]

**Pull request merged**